Windows Fix

gysddn opened the pull request:
This will solve the Windows port issues:

- [x] Using `io/fs` relative paths, gradually updating the `filepath` calls
- [x] Handling user input paths on Windows, i.e `env86 boot E:\MyVMs\alpine`
- [x] Updating home directory
- [x] Fixing `Makefile` problems
   - [x] `GOOS` and `GOARCH` env variables
   - [x] Problem with assets directory for docker, has to be passed as `.\assets:/dst`

gysddn committed:
Add fsutil functions for native to fs relative path converions

- Add RootFS to get the current root of the filesystem, this will
be always "/" for *nix system and the drive path for Windows.

- Add RootFSRelativePath function to convert a native os path
to a relative to root (/ or C:) path

gysddn committed:
Update Makefile to work on windows

- Pass environment variables to 'guest' target through make's own
'export' directive
- Pass assets dir in a OS native way for windows, otherwise docker
will just ignore it

gysddn committed:
Use APPDATA instead of user home for env86/ dir on Windows

progrium commented:
Looks good so far, but should I wait until the last bits are done? Just going by the checklist in the PR description.

gysddn commented:
So far, only the happy path is working. This should enable a user to "pull" an image and "boot" it, but other parts were on hold. I was waiting for an input to continue working on this.

progrium commented:
Okay, can you make a link for any checklist items that have a separate issue? If I missed something, let me know what you need input on.

gysddn commented:
It's only these two that are not done:

- Using io/fs relative paths, gradually updating the filepath calls (have to go through whole project and update them all)
- Handling user input paths on Windows, i.e env86 boot E:\MyVMs\alpine (whenever there is a user input)

Tara told me that only "happy path" was necessary, so I fixed that initial stage and waited for an input on where to go from there.

I just saw your comment on #8 and that was the other one, and I'll leave that one. I believe that is a filesystem problem, because even if the extraction was successful with `GNU tar` all the links were broken.

progrium commented:
Okay. Those last two seem like they would be good to do before we merge and test again, so you're not blocked to do them?

gysddn commented:
I'm not. I can start working on them and update the PR 👍🏼

gysddn committed:
Update image related path operations

- Updated image related path operations to use "path" instead of
"path/filepath"
- some `namespacefs` and `fsutil` functions are also updated to
make image work

This commit finishes the following files' transitions:
- image.go
- fsutil/copy.go

gysddn committed:
Update namespacefs to use path

- Updated namespacefs operations to use "path" instead of "filepath"
These operations have been updated and tested(where possible):

- Base()
- Clean()
- Dir()
- Join()

gysddn committed:
Fix prepare.go on windows

- Update prepare.go to pass new cross platform fsutil functions
to `DirExists`

gysddn committed:
Update pull command's path operations

- Update pull's path operation to be cross platform
- Remove 'sh' call and use osfs for copy operation
- document globalImage for Windows

gysddn committed:
Update pull cmd's path argument

- Fixed the second argument of pull when it's a relative non-existent
path for Windows.

gysddn committed:
Update boot cmd's path argument

- Fixed the argument of boot when it's a relative path, for Windows.

gysddn committed:
Update prepare cmd's path argument

- Fixed the argument of prepare when it's a relative path, for Windows.

gysddn committed:
Update create cmd's exists call for Windows

gysddn committed:
Update IsDir and Exists calls in create cmd

gysddn commented:
Input paths are handled, these files have been updated:

- `pull.go`
- `boot.go`
- `create.go`
- `prepare.go`

These files either handle input paths correctly or don't have paths as inputs:
- `assets.go` → Input is not a path.
- `network.go` → No arguments used.
- `run.go` → Immediately calls `LoadImage`, which handles paths correctly. (But this gives "Guest not available error", it seems to handle paths properly but command itself is not working, or I'm not doing it right.)
- `serve.go` → Immediately calls `LoadImage`, which handles paths correctly.

gysddn committed:
update Makefile to generate .exe extension file on Windows

gysddn commented:
Only these files left:
- `create.go`: It works and does produce an output with "dir" arg, but "docker" argument and subsequently the "tar" problem stays here, so could not test it thoroughly.
- `v86util.go`: This is only used in `create.go`, so can't test it yet.
- `tarfs/file.go` and `tarfs/fs.go`: Same problem, related to "tar" so cannot test it yet. But usage of "filepath" is very minimal here, only used to split and join paths, so change would not be big at all. This can be handled with "tar" problem.

gysddn commented:
Another thing to report here is the `osfs` module from `toolkit-go`.

The object returned by `osfs.New()` is just an empty object and all implemented `io/fs` calls are wrappers around `os` functions, i.e. `osfs.FS.Stat()` calls the `os.Stat()` with a full OS native path. Now, this works, because paths that are passed to these specific FS object are paths that are, in fact, expected to be OS native paths. So at the current state, paths that are calculated using `path/filepath` do work here and "have to" be OS native paths. But I guess, breaks the rule of go's `fs` implementation, and I'm not sure where or if this would break somewhere.

One usage is here:
https://github.com/progrium/env86/blob/d2f218fb1ba26e4327774a02deab1837247456b0/assets/assets.go#L28

```go
	if err := fsutil.CopyFS(Dir, "env86.js", osfs.New(), filepath.Join(tmpDir, "env86.js")); err != nil {
		return nil, err
	}
	if err := fsutil.CopyFS(Dir, "duplex.min.js", osfs.New(), filepath.Join(tmpDir, "duplex.min.js")); err != nil {
		return nil, err
	}
```

gysddn commented:
Other than these, the PR is ready for review.

gysddn committed:
Delete the debug binary file (previous mistake)
